safehttp/plugins/fetch_metadata/fetch_metadata_test.go

```diff
@@ -0,0 +1,399 @@
+				req.Header.Add("Sec-Fetch-Mode", "navigate")
+				req.Header.Add("Sec-Fetch-Dest", "image")
+				return req
+			}(),
+			wantStatus:  safehttp.StatusOK,
+			wantHeaders: map[string][]string{},
+			wantBody:    "",
+			wantReport:  "",
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			p := fetchmetadata.NewPlugin()
+			logger := &fooLog{}
+			p.SetReportOnly(logger)
+			rec := safehttptest.NewResponseRecorder()
+			p.Before(rec.ResponseWriter, test.req)
+
+			if status := safehttp.StatusCode(rec.Status()); status != test.wantStatus {
+				t.Errorf("status code got: %v want: %v", status, test.wantStatus)
+			}
+			if diff := cmp.Diff(test.wantHeaders, map[string][]string(rec.Header())); diff != "" {
+				t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+			}
+			if got := rec.Body(); got != test.wantBody {
+				t.Errorf("response body got: %q want: %q", got, test.wantBody)
+			}
+			if logger.report != test.wantReport {
+				t.Errorf("logger.report: got %s, want %s", logger.report, test.wantReport)
+			}
+		})
+	}
+}
+
+func TestReportModeMissingLogger(t *testing.T) {
+	p := fetchmetadata.NewPlugin()
+	defer func() {
+		if r := recover(); r == nil {
+			t.Error("p.SetReportOnly(nil) expected panic")
+		}
+	}()
+	p.SetReportOnly(nil)
+}
+
+func TestChangeEnforceReportMode(t *testing.T) {
+	logger := &fooLog{}
+	p := fetchmetadata.NewPlugin()
+	req := safehttptest.NewRequest("POST", "https://spaghetti.com/pizza", nil)
+	req.Header.Add("Sec-Fetch-Site", "cross-site")
+	req.Header.Add("Sec-Fetch-Mode", "navigate")
+	req.Header.Add("Sec-Fetch-Dest", "image")
+	rec := safehttptest.NewResponseRecorder()
+
+	p.Before(rec.ResponseWriter, req)
+
+	if want, got := safehttp.StatusForbidden, safehttp.StatusCode(rec.Status()); want != got {
+		t.Errorf("status code got: %v want: %v", got, want)
+	}
+	wantHeaders := map[string][]string{
+		"Content-Type":           {"text/plain; charset=utf-8"},
+		"X-Content-Type-Options": {"nosniff"},
+	}
+	if diff := cmp.Diff(wantHeaders, map[string][]string(rec.Header())); diff != "" {
+		t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+	}
+	if want, got := "Forbidden\n", rec.Body(); got != want {
+		t.Errorf("response body got: %q want: %q", got, want)
+	}
+
+	rec = safehttptest.NewResponseRecorder()
+	p.SetReportOnly(logger)
+
+	p.Before(rec.ResponseWriter, req)
+
+	if want, got := safehttp.StatusOK, safehttp.StatusCode(rec.Status()); want != got {
+		t.Errorf("status code got: %v want: %v", got, want)
+	}
+	if diff := cmp.Diff(map[string][]string{}, map[string][]string(rec.Header())); diff != "" {
+		t.Errorf("rec.Header() mismatch (-want +got):\n%s", diff)
+	}
+	if want, got := "", rec.Body(); got != want {
+		t.Errorf("response body got: %q want: %q", got, want)
+	}
+	if want := "POST /pizza"; logger.report != want {
+		t.Errorf("logger.report: got %s, want %s", logger.report, want)
+	}
+}
+
+func TestEnableDisableNavIsolation(t *testing.T) {
```

Comment:
In a real use case we would never get a request and then change from `NavIsolation` to not `NavIsolation` between requests. The plugins will be configured at start up and then run with the same configuration for the entire runtime of the server. So make a test that just has `NavIsolation` on from the start and don't change it I would say. Same principle with `TestChangeEnforceReportMode`.